kb/how-to/add-test-generation.md

```diff
@@ -0,0 +1,478 @@
+To add test cases for a function, a concrete class derived from `BaseTarget` must be
+added to the script.
+This class must implement all attributes and methods of `BaseTarget` and will generate
+test cases.
+This is a [test function class](#test-function-classes).
+
+As an example, test case generation is added for the test function `mpi_add_mpi()`.
+The added class will be derived from `BignumTarget`, and can be added either to
+[tests/scripts/generate_bignum_tests.py](https://github.com/Mbed-TLS/mbedtls/blob/development/tests/scripts/generate_bignum_tests.py)
+or the example script described in the previous section.
+
+### Creating the class
+
+The class must derive from a file target class, and the following class attributes should
+be set:
+ - `count = 0`: this resets the test case counting mechanism.
+Alternatively, to disable test case numbering in descriptions, set
+`show_test_count = False`.
+ - `test_function`: the test function that the class will generate cases for.
+ - `test_name`: a short descriptive name for the test, this can be the same as
+`test_function` or a more readable equivalent.
+This is used as the start of the description line.
+
+For the example, add the following class:
+
+```python
+class BignumAddExample(BignumTarget):
+    """Test cases for bignum addition."""
+    count = 0
+    test_function = "mpi_add_mpi"
+    test_name = "MPI add"
+```
+
+### Class initialization
+
+Each instance of the class will represent a different test case, each with unique inputs.
+To initialize an instance of the class, these inputs must be passed to the `__init__()`
+constructor method.
+
+For `mpi_add_mpi()`, two input values are required, `val_a` and `val_b`.
+As the test function uses string inputs, the constructor will use the same format and
+store the values as-is.
+This ensures the values are correct in the generated data, preserving leading zeros and
+null strings.
+For calculating the expected result, these inputs are also converted to, and stored in,
+`int` form; this is a multi-precision type in Python and so can be used for bignum
+calculations.
+
+For the example, add the following constructor to `BignumAddExample`:
+
+```python
+    def __init__(self, val_a: str, val_b: str) -> None:
+        self.arg_a = val_a
+        self.arg_b = val_b
+        # int(val, 16) converts from a hex string to int
+        # empty strings should be treated as zero in calculations
+        self.int_a = int(val_a, 16) if val_a else 0
+        self.int_b = int(val_b, 16) if val_b else 0
+```
+
+### Implementing abstract BaseTarget methods
+
+Abstract methods declared in `BaseTarget` must also be implemented in test function
+classes.
+
+For the example, the following imports will be required:
+
+```python
+import itertools
+from typing import Iterator, List
+# After import scripts_path
+from mbedtls_dev import test_case
+```
+
+`arguments()` returns the list of arguments required by the test function.
+Any C-string arguments must be quoted.
```

Comment:
I'm not sure what “C-string arguments” refers to here. Is this about test function arguments that are passed to test functions as `char *`? If so this statement is incomplete: `data_t` arguments also need to be quoted. It would be clearer to formulate this differently, something like

> Arguments must have the syntax expected in `.data` files, with double quotes around arguments that are strings or hex data.